### PR TITLE
DAOS-2649 tests: Adding nvme_io_verification.py test

### PR DESCRIPTION
**Pull request by saurabhtandan**
As part of this patch two new files have been added:
ftest/io/nvme_io_verification.py
ftest/io/nvme_io_verification.yaml

Some modifications have been made to ior_utils and
ior_test_base.py to accomodate some changes.

### CODE REVIEW

File: src/tests/ftest/util/agent_utils.py [hunk 1]
```diff
@@ -68,6 +68,8 @@ def run_agent(basepath, server_list, client_list=None):
     # if empty client list, 'self' is effectively client
     if client_list is None:
         client_list = [socket.gethostname().split('.', 1)[0]]
+    elif socket.gethostname().split('.', 1)[0] not in client_list:
+        client_list += [socket.gethostname().split('.', 1)[0]]
```
Comment (ravalsam) [inline src/tests/ftest/util/agent_utils.py]: This should not be part of this PR.

File: src/tests/ftest/util/agent_utils.py [hunk 2]
```diff
@@ -158,6 +160,8 @@ def stop_agent(sessions, client_list=None):
...
```
Comment (ravalsam) [inline src/tests/ftest/util/agent_utils.py]: Should not be part of this PR

File: src/tests/ftest/io/nvme_io_verification.yaml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ createmode:
+  mode: 511
+ createset:
+  setname: daos_server
+ createsvc:
+  svcn: 1
+ior:
+ client_processes:
+   np: 1
+ flags: "-w -r -k -vv"
```
Comment (ravalsam) [inline src/tests/ftest/io/nvme_io_verification.yaml]: Can you add the actual IO verification as well -W -R?

File: src/tests/ftest/io/nvme_io_verification.yaml [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# change host names to your reserved nodes, the
+# required quantity is indicated by the placeholders
+
+hosts:
+ test_servers:
+   - boro-A
```
Comment (ravalsam) [inline src/tests/ftest/io/nvme_io_verification.yaml]: Can we use full set of nodes reserved by CI?

File: src/tests/ftest/io/nvme_io_verification.yaml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+pool:
+ createmode:
+  mode: 511
+ createset:
+  setname: daos_server
+ createsvc:
+  svcn: 1
+ior:
+ client_processes:
+   np: 1
```
Comment (ravalsam) [inline src/tests/ftest/io/nvme_io_verification.yaml]: Can we use maximum CN process instead of just 1?

File: src/tests/ftest/util/ior_test_base.py [hunk 6]
```diff
@@ -102,13 +102,15 @@ def get_job_manager_command(self):
...
-    def run_ior(self, manager, processes):
+    def run_ior(self, manager, processes=None):
         """Run the IOR command.
 
         Args:
             manager (str): mpi job manager command
             processes (int): number of host processes
         """
+        if self.processes and processes is None:
+            processes = self.processes
```
Comment (phender) [inline src/tests/ftest/util/ior_test_base.py]: Why wouldn't we just pass in self.processes into this method?

File: src/tests/ftest/util/ior_test_base.py [hunk 7]
```diff
@@ -125,8 +127,12 @@ def verify_pool_size(self, original_pool_info, processes):
             original_pool_info (PoolInfo): Pool info prior to IOR
             processes (int): number of processes
         """
+        if isinstance(self.pool, TestPool):
+            self.pool = self.pool.pool
+
         # Get the current pool size for comparison
-        current_pool_info = self.pool.pool.pool_query()
+        # pylint: disable=no-member
+        current_pool_info = self.pool.pool_query()
```
Comment (phender) [inline src/tests/ftest/util/ior_test_base.py]: It is not a good idea to redefine a self attribute as a different object type.   Assuming the issue here is you want to support self.pool being assigned as either a DaosPool or Test Pool object, then I'd recommend the following:
```
if isinstance(self.pool, TestPool):
    self.pool.get_info()
    current_pool_info = self.pool.info
elif isinstance(self.pool, DaosPool):
    current_pool_info = self.pool.pool_query()
else:
    raise Exception("Unsupported pool object class: {}".format(type(self.pool)))
```

Alternatively you could just pick one object type and require that self.pool is an object of that type.

File: src/tests/ftest/util/ior_test_base.py (hunk 6 above)
Comment (saurabhtandan) [reply src/tests/ftest/util/ior_test_base.py]: We can't do that as nvme_io.py is using different set of processes for different test case. It will break that test.

File: src/tests/ftest/util/ior_test_base.py (hunk 7 above)
Comment (saurabhtandan) [reply src/tests/ftest/util/ior_test_base.py]: Will do.

File: src/tests/ftest/util/ior_utils.py [hunk 8]
```diff
@@ -147,7 +147,7 @@ def set_daos_pool_params(self, pool, display=True):
             display (bool, optional): print updated params. Defaults to True.
         """
         self.daos_pool.update(
-            pool.pool.get_uuid_str(), "daos_pool" if display else None)
+            pool.get_uuid_str(), "daos_pool" if display else None)
```
Comment (phender) [inline src/tests/ftest/util/ior_utils.py]: This change will break other tests that use `ior_utils.py` and call `set_daos_params()` with a `pool` argument that is a `TestPool` type, for example:
- container_create.py
- metadata.py
- ior_test_base,py
- mdtest_test_base.py

Comment (saurabhtandan) [reply src/tests/ftest/util/ior_utils.py]: ior_test_base.py and mdtest_test_base.py did ran fine as part of CI run but to address the other two I have moved from using DaosPool to TestPool.

File: src/tests/ftest/io/nvme_io_verification.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+                # Verify IOR consumed the expected amount ofrom the pool
+                self.verify_pool_size(size_before_ior, self.processes)
+
+            try:
+                if self.pool:
+                    # pylint: disable=attribute-defined-outside-init
+                    self.pool.destroy(1)
+                    self.pool = None
+            except DaosApiError as error:
```
Comment (phender) [inline src/tests/ftest/io/nvme_io_verification.py]: The `TestPool.destroy()` method catches the `DaosApiError` exception and instead returns an `avocado.core.exceptions.TestFail` exception.  Here you want to catch the `TestFail` exception.

File: src/tests/ftest/util/ior_test_base.py [hunk 10]
```diff
@@ -78,6 +79,7 @@ def tearDown(self):
     def create_pool(self):
         """Create a TestPool object to use with ior."""
         # Get the pool params
+        # pylint: disable=attribute-defined-outside-init
```
Comment (phender) [inline src/tests/ftest/util/ior_test_base.py]: This particular error is a result of pylint not being able to find `apricot/test.py` and realize that `TestWithServers.__init__()` defines `self.pool`.  With DAOS-3731 merged, this should hopefully not be a problem and this disable option should be removed.

File: src/tests/ftest/util/ior_test_base.py [hunk 11]
```diff
@@ -202,7 +204,15 @@ def verify_pool_size(self, original_pool_info, processes):
             processes (int): number of processes
         """
         # Get the current pool size for comparison
-        current_pool_info = self.pool.pool.pool_query()
+        if isinstance(self.pool, TestPool):
+            self.pool.get_info()
+            current_pool_info = self.pool.info
+        elif isinstance(self.pool, DaosPool):
```
Comment (phender) [inline src/tests/ftest/util/ior_test_base.py]: Since defining `self.pool` as a `DaosPool` object will cause other problems in this class I don't think it makes sense to support the alternate definition here.  If you agree, then the addition on line 33 can also be removed.

Comment (saurabhtandan) [reply src/tests/ftest/util/ior_test_base.py]: As of nvme_io is using DaosPool which is also using this method. So until we move nvme_io to use TestPool as well I think we should keep this. In the changes for nvme_io when it is moved from using DaosPool to TestPool, we can remove the support for DaosPool from this method as well.

File: src/tests/ftest/util/ior_test_base.py (hunk 10 above)
Comment (saurabhtandan) [reply src/tests/ftest/util/ior_test_base.py]: Will do.

File: src/tests/ftest/io/nvme_io_verification.py (hunk 9 above)
Comment (saurabhtandan) [reply src/tests/ftest/io/nvme_io_verification.py]: Will do
